### PR TITLE
pkcs11 symmetric ciphers

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

File: ta/pkcs11/src/processing.c [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+ * entry_processing_step - Generic entry on active processing
+ *
+ * @ctrl = [session-handle]
+ * @ptype = Invocation parameter types
+ * @params = Invocation parameters reference
+ * @function - encrypt, decrypt, sign, verify, digest, ...
+ * @step - update, oneshot, final
+ *
+ * The generic part come that all the commands uses the same
+ * input/output invocation parameters format (ctrl/in/out).
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing.c:221]: ditto

File: ta/pkcs11/src/processing.c [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+	TEE_Free(proc_params);
+
+	return rc;
+}
+
+/*
+ * entry_processing_step - Generic entry on active processing
+ *
+ * @ctrl = [session-handle]
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing.c:214]: ditto

File: ta/pkcs11/src/processing.c [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+		break;
+	}
+
+	return rc;
+}
+
+/*
+ * entry_verify_oneshot - Generic entry on active processing
+ *
+ * @ctrl = [session-handle]
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing.c:289]: ditto

File: ta/pkcs11/src/processing.c [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+ * @ptype = Invocation parameter types
+ * @params = Invocation parameters reference
+ * @function - encrypt, decrypt, sign, verify, digest, ...
+ * @step - update, oneshot, final
+ *
+ * The generic part come that all the commands uses the same
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing.c:295]: ditto

File: ta/pkcs11/src/processing.c [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+		if (rc != PKCS11_CKR_BUFFER_TOO_SMALL)
+			release_active_processing(session);
+		break;
+	}
+
+	return rc;
+}
+
+/*
+ * entry_verify_oneshot - Generic entry on active processing
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing.c:287]:
```patch
- * entry_verify_oneshot - Generic entry on active processing
+ * entry_verify_oneshot - Run a single part verification processing

File: ta/pkcs11/src/processing.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2017-2020, Linaro Limited
+ */
+
+#ifndef PKCS11_TA_PROCESSING_H
+#define PKCS11_TA_PROCESSING_H
+
+#include <tee_internal_api.h>
+#include <pkcs11_attributes.h>
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing.h:10]: swap the 2 lines

File: ta/pkcs11/src/processing_aes.c [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#include <assert.h>
+#include <compiler.h>
+#include <tee_internal_api.h>
+#include <trace.h>
+#include <util.h>
+
+#include "pkcs11_token.h"
+#include "processing.h"
+#include "serializer.h"
+#include "pkcs11_helpers.h"
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:15]: move at line 12.

File: ta/pkcs11/src/processing_aes.c [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+enum pkcs11_rc tee_init_ctr_operation(struct active_processing *processing,
+				      void *proc_params, size_t params_size)
+{
+	struct serialargs args;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:20]: prefer `struct serialargs args = { };` and remove `TEE_MemFill()` below.

File: ta/pkcs11/src/processing_symm.c [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+#include <tee_api_defines.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+#include <utee_defines.h>
+#include <util.h>
+
+#include "attributes.h"
+#include "object.h"
+#include "pkcs11_token.h"
+#include "pkcs11_attributes.h"
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:18]: few reordering...

File: ta/pkcs11/src/processing_symm.c [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+	if (res == TEE_ERROR_NOT_SUPPORTED)
+		return PKCS11_CKR_MECHANISM_INVALID;
+
+	return tee2pkcs_error(res);
+}
+
+static enum pkcs11_rc load_tee_key(struct pkcs11_session *session,
+				   struct pkcs11_object *obj)
+{
+	TEE_Attribute tee_attr;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:129]: `= { };` and remove `TEE_MemFill()`.

File: ta/pkcs11/src/processing_symm.c [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+
+key_ready:
+	res = TEE_SetOperationKey(session->processing->tee_op_handle,
+				  obj->key_handle);
+	if (res) {
+		DMSG("TEE_SetOperationKey failed, %#"PRIx32, res);
+		goto error;
+	}
+
+	return tee2pkcs_error(res);
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:177]: minor: I think returning `PKCS11_CKR_OK` straight would be better for maintenance.

File: ta/pkcs11/src/entry.c [hunk 12]
```diff
@@ -208,6 +209,40 @@ TEE_Result TA_InvokeCommandEntryPoint(void *tee_session, uint32_t cmd,
...
+		break;
+	case PKCS11_CMD_DECRYPT_UPDATE:
+		rc = entry_processing_step(client, ptypes, params,
+					   PKCS11_FUNCTION_DECRYPT,
+					   PKCS11_FUNC_STEP_UPDATE);
+		break;
+	case PKCS11_CMD_ENCRYPT_ONESHOT:
+		rc = entry_processing_step(client, ptypes, params,
+					   PKCS11_FUNCTION_ENCRYPT,
+					   PKCS11_FUNC_STEP_ONESHOT);
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply ta/pkcs11/src/entry.c:233]: Thanks, I'll fix.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 13]
```diff
@@ -24,6 +24,95 @@
...
+	case PKCS11_FUNCTION_SIGN:
+		return PKCS11_CKFM_SIGN;
+	case PKCS11_FUNCTION_VERIFY:
+		return PKCS11_CKFM_VERIFY;
+	case PKCS11_FUNCTION_SIGN_RECOVER:
+		return PKCS11_CKFM_SIGN_RECOVER;
+	case PKCS11_FUNCTION_VERIFY_RECOVER:
+		return PKCS11_CKFM_VERIFY_RECOVER;
+	default:
+		return 0;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply ta/pkcs11/src/pkcs11_attributes.c:55]: `0` doesn't seem to be a valid value. This is a flags field where I assume that at least one flag is expected to be set.
There's even a test for that in `check_mechanism_against_processing()` below.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 14]
```diff
@@ -773,3 +892,178 @@ enum pkcs11_rc check_created_attrs_against_processing(uint32_t proc_id,
...
+	if (function == PKCS11_FUNCTION_UNWRAP &&
+	    !get_bool(head, PKCS11_CKA_UNWRAP)) {
+		DMSG("unwrap not permitted");
+		return PKCS11_CKR_KEY_FUNCTION_NOT_PERMITTED;
+	}
+	if (function == PKCS11_FUNCTION_DERIVE &&
+	    !get_bool(head, PKCS11_CKA_DERIVE)) {
+		DMSG("derive not permitted");
+		return PKCS11_CKR_KEY_FUNCTION_NOT_PERMITTED;
+	}
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply ta/pkcs11/src/pkcs11_attributes.c:1038]: I'll try something.

File: ta/pkcs11/src/pkcs11_attributes.h [hunk 15]
```diff
@@ -134,4 +134,30 @@ enum pkcs11_rc check_created_attrs_against_token(struct pkcs11_session *session,
...
+enum pkcs11_rc check_created_attrs(struct obj_attrs *key1,
+				   struct obj_attrs *key2);
+
+/*
+ * Check the attributes of the parent secret (key) used in the processing
+ * do match the target processing.
+ *
+ * @proc_id - PKCS11_CKM_xxx
+ * @subproc_id - boolean attribute encrypt or decrypt or sign or verify, if
+ *		 applicable to proc_id.
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply ta/pkcs11/src/pkcs11_attributes.h:146]: I'll keep an updated inline comment.

File: ta/pkcs11/src/pkcs11_helpers.h [hunk 16]
```diff
@@ -10,6 +10,9 @@
 #include <stddef.h>
 
 #include <token_capabilities.h>
+#include <pkcs11_attributes.h>
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply ta/pkcs11/src/pkcs11_helpers.h:13]: OK

**Comment by jenswi-linaro** [comment]
Addressing comments

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply ta/pkcs11/src/pkcs11_attributes.c:55]: You're right, sorry, these are bit flags.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 17]
```diff
@@ -24,6 +24,95 @@
...
+		switch (function) {
+		case PKCS11_FUNCTION_IMPORT:
+		case PKCS11_FUNCTION_COPY:
+		case PKCS11_FUNCTION_MODIFY:
+		case PKCS11_FUNCTION_DESTROY:
+			return PKCS11_CKR_OK;
+		default:
+			break;
+		}
+		allowed = mechanism_supported_flags(mechanism_type) &&
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:78]: I think here it should be `&`, not `&&`: mechanism supports the target function.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 18]
```diff
@@ -690,6 +779,36 @@ static enum pkcs11_rc check_attrs_misc_integrity(struct obj_attrs *head)
...
+						struct obj_attrs *head)
+{
+	bool private = true;
+
+	switch (get_class(head)) {
+	case PKCS11_CKO_SECRET_KEY:
+	case PKCS11_CKO_PUBLIC_KEY:
+	case PKCS11_CKO_DATA:
+		if (!get_bool(head, PKCS11_CKA_PRIVATE))
+			private = false;
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:795]: minor: could be __simplified__ to `private = get_bool(head, PKCS11_CKA_PRIVATE);`

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 19]
```diff
@@ -773,3 +892,170 @@ enum pkcs11_rc check_created_attrs_against_processing(uint32_t proc_id,
...
+		EMSG("Length %"PRIu32" vs range [%"PRIu32" %"PRIu32"]",
+		     key_length, min_key_size, max_key_size);
+
+		return PKCS11_CKR_KEY_SIZE_RANGE;
+	}
+
+	return PKCS11_CKR_OK;
+}
+
+/* Check processing ID against attribute ALLOWED_PROCESSINGS if any */
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:968]: s/ALLOWED_PROCESSINGS/ALLOWED_MECHANISMS/

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply ta/pkcs11/src/pkcs11_attributes.c:1030]: nice!

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 20]
```diff
@@ -773,3 +892,170 @@ enum pkcs11_rc check_created_attrs_against_processing(uint32_t proc_id,
...
+
+		if (proc == proc_id)
+			return true;
+	}
+
+	DMSG("can't find %s in allowed list", id2str_proc(proc_id));
+	return false;
+}
+
+static uint32_t func_to_attr(enum processing_func func)
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:995]: maybe replace `uint32_t` with `enum pkcs11_attr_id`.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 21]
```diff
@@ -773,3 +892,170 @@ enum pkcs11_rc check_created_attrs_against_processing(uint32_t proc_id,
...
+	enum pkcs11_class_id key_class = get_class(head);
+	enum pkcs11_key_type key_type = get_key_type(head);
+	uint32_t attr = func_to_attr(function);
+
+	if (!get_bool(head, attr)) {
+		DMSG("%s not permitted", id2str_attr(attr));
+		return PKCS11_CKR_KEY_FUNCTION_NOT_PERMITTED;
+	}
+
+	/* Check processing complies for parent key family */
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:1032]: `... complies for ...` is ok? 
maybe `... complies with ...`.

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 17 above)
Comment (jenswi-linaro) [reply ta/pkcs11/src/pkcs11_attributes.c:78]: Agree.

**Comment by jenswi-linaro** [comment]
Comments addressed.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 22]
```diff
@@ -773,3 +896,170 @@ enum pkcs11_rc check_created_attrs_against_processing(uint32_t proc_id,
...
+		return PKCS11_CKA_UNDEFINED_ID;
+	}
+}
+
+enum pkcs11_rc
+check_parent_attrs_against_processing(enum pkcs11_mechanism_id proc_id,
+				      enum processing_func function,
+				      struct obj_attrs *head)
+{
+	enum pkcs11_rc __maybe_unused rc = PKCS11_CKR_OK;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply ta/pkcs11/src/pkcs11_attributes.c:1026]: Thanks, I'll remove it.

File: ta/pkcs11/src/token_capabilities.c [hunk 23]
```diff
@@ -118,12 +133,39 @@ bool __maybe_unused mechanism_flags_complies_pkcs11(uint32_t mechanism_type,
...
 const struct pkcs11_mechachism_modes token_mechanism[] = {
-	MECHANISM(PKCS11_CKM_AES_ECB, 0, 0),
+	TA_MECHANISM(PKCS11_CKM_AES_ECB, CKFM_CIPHER),
+	TA_MECHANISM(PKCS11_CKM_AES_CBC, CKFM_CIPHER),
+	TA_MECHANISM(PKCS11_CKM_AES_CBC_PAD, CKFM_CIPHER),
+	TA_MECHANISM(PKCS11_CKM_AES_CTR, CKFM_CIPHER),
+	TA_MECHANISM(PKCS11_CKM_AES_CTS, CKFM_CIPHER),
+	TA_MECHANISM(PKCS11_CKM_AES_ECB_ENCRYPT_DATA, PKCS11_CKFM_DERIVE),
+	TA_MECHANISM(PKCS11_CKM_AES_CBC_ENCRYPT_DATA, PKCS11_CKFM_DERIVE),
+	TA_MECHANISM(PKCS11_CKM_AES_KEY_GEN, PKCS11_CKFM_GENERATE),
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply ta/pkcs11/src/token_capabilities.c:168]: OK, I'll remove it.

**Comment by jenswi-linaro** [comment]
Comments addressed

**Comment by jenswi-linaro** [comment]
Fixed the travis error, it will go away completely when squashing.

**Comment by vesajaaskelainen** [comment]
Did lightweight review for the changes in here. Didn't spot anything with this.

I would say one would go to quite deep in implementation details to notice further possible problems.

Are you looking for that kind of review? -- I believe that would require a bit more time to go thru the internal logic and in reality probably when we are symmetric ready then test suite probably would covert most of those cases anyway.

**Comment by jenswi-linaro** [comment]
Thanks, I think lightweight is the way to go.

**Comment by vesajaaskelainen** [comment]
`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

**Comment by jenswi-linaro** [comment]
Tags applied.
Note that I added:
```
    Co-developed-by: Etienne Carriere <etienne.carriere@linaro.org>
    Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>
```
too, i missed them earlier. @etienne-lms you need to shout if that's not OK.

**Review by etienne-lms — APPROVED** [review]
looks all good to me.

**Pull request merged**